intro_to_ada.rst

```diff
@@ -2681,21 +2681,1055 @@ Protected objects: entries
+
+Similar to subprograms, we could use the ``External_Name`` aspect to
+rename the variable in the Ada application.
+
+Using Ada variables in C
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is also possible to use variables declared in Ada files in C
+applications. Similarly to subprogram, this requires the use of the
+``Export`` aspect.
+
+Let's reuse the previous example and add a counter, as we did in the
+previous example:
+
+.. code-block:: ada
+
+    --% filename: c_api.ads
+
+    with Interfaces.C;
+    use  Interfaces.C;
+
+    package C_API is
+
+       func_cnt : int := 0
+         with
+           Export     => True,
+           Convention => C;
+
+       function My_Func (a : int) return int
+         with
+           Export        => True,
+           Convention    => C,
+           External_Name => "my_func";
+
+    end C_API;
+
+The variable is then increment in ``My_Func``:
+
+.. code-block:: ada
+
+    --% filename: c_api.adb
+    package body C_API is
+
+       function My_Func (a : int) return int is
+       begin
+          func_cnt := func_cnt + 1;
+          return a * 2;
+       end My_Func;
+
+    end C_API;
+
+In the C application, we just need to declare the variable and use it:
+
+.. code-block:: c
+
+    #include <stdio.h>
+
+    extern int my_func (int a);
+
+    extern int func_cnt;
+
+    int main (int argc, char **argv) {
+
+      int v;
+
+      v = my_func(1);
+      v = my_func(2);
+      v = my_func(3);
+
+      printf("Result is %d\n", v);
+
+      printf("Function was called %d times\n", func_cnt);
+
+      return 0;
+    }
+
+Again, by running the application, we see that the value from the counter
+will contain the correct number of times that ``my_func`` was called.
+
+Using unconstrained types
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the previous examples, we're being careful about the data types: all of
+them are coming from the ``Interfaces.C`` package. Using Ada built-in
+types when interfacing with C can be problematic, especially in case of
+unconstrained types. For example:
+
+.. code-block:: c
+
+    /*% filename: test.h */
+
+    char * my_func (void);
+
+This is the function implementation:
+
+.. code-block:: c
+
+    #include <stdio.h>
+    #include "test.h"
+
+    char * my_func (void)
+    {
+      return "hello";
+    }
+
+In the Ada application, we try to import this as a ``String`` type:
+
+.. code-block:: ada
+
+    with Interfaces.C;
+    use  Interfaces.C;
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    procedure Show_C_Func is
+
+       function my_func return String
+         with
+           Import        => True,
+           Convention    => C;
+
+       S : String := my_func;
+
+    begin
+       Put_Line (S);
+    end Show_C_Func;
+
+When running this application, we'll get a ``STORAGE_ERROR`` exception.
+Therefore, the recommendation is to be very careful about the data types
+and use the ``Interfaces.C`` package whenever possible for interfacing
+with C.
+
 Multi-language project
```

Comment:
This should come before, almost at the beginning in my opinion, so that people can start experimenting very early.

Reply:
Just FYI: it was actually you who originally put the "Multi-language project" section at the end. ;-) 

Anyway, I agree with you that this could be at the beginning.